python/ray/experimental/state/api.py

```diff
@@ -43,37 +45,96 @@
 
 
 class StateApiClient(SubmissionClient):
-    """State API Client issues REST GET requests to the server for resource states.
-
-    Args:
-        api_server_address: The address of API server. If it is not give, it assumes
-        the ray is already connected and obtains the API server address using Ray API.
-    """
+    """State API Client issues REST GET requests to the server for resource states."""
 
     def __init__(
         self,
-        api_server_address: str = None,
+        address: Optional[str] = None,
+        cookies: Optional[Dict[str, Any]] = None,
+        headers: Optional[Dict[str, Any]] = None,
     ):
+        """Initialize a StateApiClient and check the connection to the cluster.
+
+        Args:
+            address: The address of Ray API server. If not provided,
+                it will be configured automatically from querying the GCS server.
+            cookies: Cookies to use when sending requests to the HTTP job server.
+            headers: Headers to use when sending requests to the HTTP job server, used
+                for cases like authentication to a remote cluster.
+        """
+        if requests is None:
+            raise RuntimeError(
+                "The Ray state CLI & SDK require the ray[default] "
+                "installation: `pip install 'ray[default']``"
+            )
+        if not headers:
+            headers = {"Content-Type": "application/json"}
         super().__init__(
-            self._get_default_api_server_address()
-            if api_server_address is None
-            else api_server_address,
+            address,
             create_cluster_if_needed=False,
-            headers={"Content-Type": "application/json"},
+            headers=headers,
+            cookies=cookies,
         )
 
-    @classmethod
-    def _get_default_api_server_address(cls) -> str:
-        assert (
-            ray.is_initialized()
-            and ray._private.worker.global_worker.node.address_info["webui_url"]
-            is not None
-        )
-        return (
-            f"http://{ray._private.worker.global_worker.node.address_info['webui_url']}"
+    def print_slow_api_warning(
+        self,
+        timeout: int,
+        cv: threading.Condition,
+        endpoint: str,
+        print_interval_s: float,
+        print_warning: bool,
+    ):
+        """Print the slow API warning. every print_interval_s seconds until timeout.
+
+        Args:
+            timeout: The API timeout.
+            cv: Condition variable. If it is notified, the thread will terminate.
+            endpoint: The endpoint to query.
+            print_interval_s: The interval to print warning message until timeout.
+            print_warning: Print the warning if it is set to True.
+        """
+        start = time.time()
+        last = time.time()
+        i = 1
+        while True:
```

Comment:
is it possible to simplify it with threading.Timer? https://docs.python.org/3/library/threading.html#threading.Timer

Reply:
I tried this first, but this requires us to have a new timer thread per 5 seconds. That says the current semantic is to print warning every 5 seconds, but if we'd like to achieve the same thing, we should create `timeout / 5` number of threads (each of which is a timer).

Alternatively, we can only have 1~2 warnings instead of every 5 seconds (a bit worse UX, but simplify the logic here).

Which one's your preference?